src/codegen/generators/variable_generator.rs

```diff
@@ -24,6 +25,11 @@ use crate::codegen::debug::DebugBuilderEnum;
 
 pub fn read_got_layout(location: &str, format: ConfigFormat) -> Result<HashMap<String, u64>, Diagnostic> {
     let mut s = String::new();
+    if !Path::new(location).is_file() {
+        // Assume if the file doesn't exist that there is no existing GOT layout yet. write_got_layout will handle
+        // creating our file when we want to.
+        return Ok(HashMap::new());
+    }
```

Comment:
this is great - I would move the `String` creation right after the check to make sure we don't allocate one for nothing, but otherwise it's solid and the check is good :)

Reply:
So something like

```rust
    if !Path::new(location).is_file() {
        // Assume if the file doesn't exist that there is no existing GOT layout yet. write_got_layout will handle
        // creating our file when we want to.
        return Ok(HashMap::new());
    }
    
    let mut fp = File::open(location).map_err(|_| Diagnostic::new("GOT layout file could not be opened"))?;
    
    let mut s = String::new();
    fp.read_to_string(&mut s).map_err(|_| Diagnostic::new("GOT layout could not be read from file"))?;
```

also, last nitpick I promise: `fp.read_to_string(&mut s)` can be replaced by [`fs::read_to_string`](https://doc.rust-lang.org/std/fs/fn.read_to_string.html) which is a helper method to do the same thing. so this would become

```rust
    if !Path::new(location).is_file() {
        // Assume if the file doesn't exist that there is no existing GOT layout yet. write_got_layout will handle
        // creating our file when we want to.
        return Ok(HashMap::new());
    }
    
    let mut fp = File::open(location).map_err(|_| Diagnostic::new("GOT layout file could not be opened"))?;
    
    let s = fs::read_to_string(&mut s).map_err(|_| Diagnostic::new("GOT layout could not be read from file"))?;
```

Reply:
Nice, thanks! Do you also thing `fs::write` is a good helper to use when we're doing the writing?